### PR TITLE
Fixing jitting with device prng key

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

When setting a prng key on the device and using jax-jit, we get a leaked tracer error.

**Description of the Change:**

If the user specifies a `PRNGKey` as a seed, we assume they want to jit end-to-end.

We raise a warning during `DefaultQubit.execute` if we are doing native jitting and have a decently sized batch. This avoids the huge compilation overhead problem.

**Benefits:**

No more leaked tracers.  Users can jit end to end by specifying a PRNGKey.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 1]
```diff
@@ -388,14 +388,20 @@ def func(x, y, z):
...
-        jaxpr = str(jax.make_jaxpr(func)(*params))
-        # will change once we solve the compilation overhead issue
-        # assert "pure_callback" not in jaxpr
-        # TODO: [sc-82874]
-        assert "pure_callback" in jaxpr
+        if diff_method == "best":
+            with pytest.warns(UserWarning, match="substantial compilation overheads"):
+                jaxpr = str(jax.make_jaxpr(func)(*params))
+        else:
+            jaxpr = str(jax.make_jaxpr(func)(*params))
```
Comment (albi3ro) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:395]:
```suggestion
        jaxpr = str(jax.make_jaxpr(func)(*params))
```

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 2]
```diff
@@ -388,14 +388,20 @@ def func(x, y, z):
...
+        assert "pure_callback" not in jaxpr
 
         func2 = jax.jit(func)
-        results2 = func2(*params)
+        if diff_method == "best":
+            # jitting parameter shift derivatives
+            with pytest.warns(UserWarning, match="substantial compilation overheads"):
+                results2 = func2(*params)
+        else:
+            results2 = func2(*params)
```
Comment (albi3ro) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:404]:
```suggestion
        results2 = func2(*params)
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -97,8 +97,8 @@
   ```
   
 * Devices can now configure whether or not ML framework data is sent to them
-  via an `ExecutionConfig.convert_to_numpy` parameter. This is not used on 
-  `default.qubit` due to compilation overheads when jitting.
+  via an `ExecutionConfig.convert_to_numpy` parameter. End-to-end jitting on
+  `default.qubit` is used if the user specified a `jax.random.PRNGKey` as a seed.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:101]:
```suggestion
  `default.qubit` is used if the user specified a `jax.random.PRNGKey` as a seed.
  [(#6899)](https://github.com/PennyLaneAI/pennylane/pull/6899)
```

**Review by JerryChen97 — APPROVED** [review]
Thanks a lot!! I tested locally and this fixes the issue of qml dev branch on `tutorial_jax_transformations.py`

File: tests/devices/default_qubit/test_default_qubit.py [hunk 4]
```diff
@@ -1486,6 +1494,34 @@ def test_finite_shots_postselection_defer_measurements(self, max_workers):
...
+        key1 = jax.random.PRNGKey(123456)
+        key2 = jax.random.PRNGKey(8877655)
+        x = jax.numpy.array(0.5)
+        # no leaked tracer errors
+        res1 = workflow(key1, x)
+        res1_again = workflow(key1, x)
+        res2 = workflow(key2, x)
+
+        assert qml.math.allclose(res1, res1_again)
+        assert not qml.math.allclose(res1, res2)
```
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit.py:1523]: This is fine for now, but just a note that JAX does not guarantee PRNG key outputs to be consistent between releases. Maybe worth adding a really low tolerance here to drive home the fact that different results are expected for different seeds.

**Review by mudit2812 — APPROVED** [review]

File: tests/devices/default_qubit/test_default_qubit.py (hunk 4 above)
Comment (JerryChen97) [inline tests/devices/default_qubit/test_default_qubit.py:1523]:
```suggestion
        assert qml.math.allclose(res1, res1_again, rtol=0.1)
        assert not qml.math.allclose(res1, res2, rtol=0.1)
```
Suggesting a personally preferred threshold for any MC test, with arguments being the std of MC are propto `1/sqrt(N)` N being the total number of trials.

Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:1523]: The chance that different keys gives the exact same raw samples would be `0.5**100 = 7e-31`.  I think we can safely say that that the samples will always be independent as long as the keys are different.

Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:1523]: We are sampling here, so `res1` and `res2` are strictly boolean integers.
